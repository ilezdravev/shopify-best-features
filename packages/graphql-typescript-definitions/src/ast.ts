import {readFileSync, readJSONSync} from 'fs-extra';
import {
  Source,
  parse,
  concatAST,
  buildClientSchema,
} from 'graphql';
import {compile} from 'graphql-tool-utilities/ast';

export interface Options {
  schemaFile: string,
  graphQLFiles: string[],
}

export default function buildContext({
  schemaFile,
  graphQLFiles,
}: Options) {
  const schemaJSON = readJSONSync(schemaFile);
  const schema = buildClientSchema(schemaJSON.data);

  const sources = graphQLFiles.map((file) => {
    const body = readFileSync(file, 'utf8');
    return new Source(body, file);
  });
  const documents = concatAST(sources.map((source) => parse(source)));
  return compile(schema, documents);
}
