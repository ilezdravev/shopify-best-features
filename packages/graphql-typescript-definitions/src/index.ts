import {writeFileSync} from 'fs-extra';
import {Operation, Fragment, AST} from 'graphql-tool-utilities/ast';

import buildAST, {Options} from './ast';
import {printFile} from './print';

export default function graphQLToTypeScriptDefinitions(options: Options) {
  const ast = buildAST(options);
  const fileMap = groupQueriesAndFragmentsByFile(ast);

  Object.keys(fileMap).forEach((path) => {
    const file = fileMap[path];

    const content = printFile(file, ast);
    if (!content) { return; }

    const newFile = `${file.path}.d.ts`;
    writeFileSync(newFile, content);
  });
}

interface File {
  path: string,
  operation?: Operation,
  fragment?: Fragment,
}

interface FileMap {
  [key: string]: File,
}

function groupQueriesAndFragmentsByFile({operations, fragments}: AST): FileMap {
  const map: FileMap = {};

  Object
    .keys(operations)
    .forEach((name) => {
      const operation = operations[name];
      map[operation.filePath] = {
        path: operation.filePath,
        operation,
      };
    });
  
  Object
    .keys(fragments)
    .forEach((name) => {
      const fragment = fragments[name];
      map[fragment.filePath] = {
        path: fragment.filePath,
        fragment,
      };
    });
  
  return map;
}
