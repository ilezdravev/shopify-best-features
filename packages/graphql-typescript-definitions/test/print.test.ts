import {buildSchema, parse, GraphQLSchema, Source, concatAST} from 'graphql';
import {compile} from 'graphql-tool-utilities/ast';

import {printFile, Options} from '../src/print';

describe('printFile()', () => {
  describe('primitive types', () => {
    it('prints a string type', () => {
      let schema = buildSchema(`
        type Query {
          name: String!
        }
      `);

      expect(print('query Name { name }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          name: String
        }
      `);

      expect(print('query Name { name }', schema)).toMatchSnapshot();
    });

    it('prints an integer type', () => {
      let schema = buildSchema(`
        type Query {
          age: Int!
        }
      `);

      expect(print('query Age { age }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          age: Int
        }
      `);

      expect(print('query Age { age }', schema)).toMatchSnapshot();
    });

    it('prints a float type', () => {
      let schema = buildSchema(`
        type Query {
          age: Float!
        }
      `);

      expect(print('query Age { age }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          age: Float
        }
      `);

      expect(print('query Age { age }', schema)).toMatchSnapshot();
    });

    it('prints a boolean type', () => {
      let schema = buildSchema(`
        type Query {
          married: Boolean!
        }
      `);

      expect(print('query Married { married }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          married: Boolean
        }
      `);

      expect(print('query Married { married }', schema)).toMatchSnapshot();
    });

    it('prints an ID type', () => {
      let schema = buildSchema(`
        type Query {
          id: ID!
        }
      `);

      expect(print('query ProductID { id }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          id: ID
        }
      `);

      expect(print('query ProductID { id }', schema)).toMatchSnapshot();
    });
  });

  describe('enum types', () => {
    it('prints the enum type and its type declaration', () => {
      let schema = buildSchema(`
        enum Episode {
          ONE
          TWO
          THREE
        }

        type Query {
          episode: Episode!
        }
      `);

      expect(print('query WhichEpisode { episode }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        enum Episode {
          ONE
          TWO
          THREE
        }

        type Query {
          episode: Episode
        }
      `);

      expect(print('query WhichEpisode { episode }', schema)).toMatchSnapshot();
    });
  });

  describe('custom scalars', () => {
    it('declares them as strings', () => {
      let schema = buildSchema(`
        scalar DateTime

        type Query {
          date: DateTime!
        }
      `);

      expect(print('query Date { date }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        scalar DateTime

        type Query {
          date: DateTime
        }
      `);

      expect(print('query Date { date }', schema)).toMatchSnapshot();
    });
  });

  describe('list types', () => {
    it('correctly wraps types that are potentially null', () => {
      let schema = buildSchema(`
        type Query {
          nicknames: [String!]!
        }
      `);

      expect(print('query Nicknames { nicknames }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          nicknames: [String]!
        }
      `);

      expect(print('query Nicknames { nicknames }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          nicknames: [String!]
        }
      `);

      expect(print('query Nicknames { nicknames }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          nicknames: [String]
        }
      `);

      expect(print('query Nicknames { nicknames }', schema)).toMatchSnapshot();
    });

    it('wraps nested lists', () => {
      let schema = buildSchema(`
        type Query {
          nicknames: [[String!]]!
        }
      `);

      expect(print('query Nicknames { nicknames }', schema)).toMatchSnapshot();

      schema = buildSchema(`
        type Query {
          nicknames: [[String]]
        }
      `);

      expect(print('query Nicknames { nicknames }', schema)).toMatchSnapshot();
    });
  });

  describe('object types', () => {
    const schema = buildSchema(`
      interface Entity {
        id: ID!
        name: String!
        nickname: String
      }

      enum Kind {
        DOG
        CAT
      }
    
      type Pet implements Entity {
        id: ID!
        name: String!
        nickname: String
        kind: Kind
      }

      type Person implements Entity {
        id: ID!
        name: String!
        nickname: String
        age: Int!
        relatives: [Person!]!
        favoritePet: Pet
      }

      type Query {
        person: Person
        entities: [Entity!]!
      }
    `);

    describe('objects fields', () => {
      it('nests them inline in the original object', () => {
        expect(
          print(
            `
          query Profile {
            person {
              name
              age
              relatives {
                name
                favoritePet {
                  name
                }
              }
            }
          }
        `,
            schema,
          ),
        ).toMatchSnapshot();
      });
    });

    describe('inline fragments', () => {
      it('treats them as their original nullability when they are guaranteed', () => {
        expect(
          print(
            `
          query Entities {
            entities {
              ...on Entity {
                name
                nickname
              }
            }
          }
        `,
            schema,
          ),
        ).toMatchSnapshot();
      });

      it('forces fields to be nullable when the type restriction is not guaranteed', () => {
        expect(
          print(
            `
          query Entities {
            entities {
              ...on Pet {
                name
                nickname
              }
            }
          }
        `,
            schema,
          ),
        ).toMatchSnapshot();
      });
    });

    describe('fragment spreads', () => {
      it('prints them as union types when the type restriction is guaranteed', () => {
        expect(
          print(
            `
          query Entities {
            entities {
              ...EntityName
            }
          }
        `,
            schema,
            {
              'subfolder/MyFragment.graphql': `
            fragment EntityName on Entity {
              name
            }
          `,
            },
          ),
        ).toMatchSnapshot();
      });

      it('prints a partial fragment type when the type restriction is not guaranteed', () => {
        expect(
          print(
            `
          query Entities {
            entities {
              ...EntityName
            }
          }
        `,
            schema,
            {
              'subfolder/MyFragment.graphql': `
            fragment EntityName on Person {
              name
            }
          `,
            },
          ),
        ).toMatchSnapshot();
      });
    });
  });

  describe('variables', () => {
    it('prints out a variable interface and all associated types', () => {
      const schema = buildSchema(`
        scalar DateTime

        enum Style {
          FULL
          NICK
        }

        input Language {
          code: String
        }

        input NameInput {
          style: Style!
          date: DateTime
          excited: Boolean!
          language: Language
        }

        type Query {
          name(input: NameInput): String!
        }
      `);

      expect(
        print(
          'query Name($myInput: NameInput) { name(input: $myInput) }',
          schema,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('mutations', () => {
    it('prints mutations', () => {
      const schema = buildSchema(`
        enum Kind {
          CAT
          DOG
        }

        type Pet {
          name: String!
          kind: Kind!
        }

        type Person {
          name: String!
          favoritePet: Pet
          pets: [Pet!]!
        }

        type Query {
          person: Person!
        }

        type Mutation {
          rename(name: String!): Person!
        }
      `);

      const document = `
        mutation Rename($name: String!) {
          rename(name: $name) {
            name
          }
        }
      `;

      expect(print(document, schema)).toMatchSnapshot();
    });
  });

  describe('fragments', () => {
    const schema = buildSchema(`
      interface Entity {
        id: ID!
        name: String!
        nickname: String
      }

      enum Kind {
        DOG
        CAT
      }
    
      type Pet implements Entity {
        id: ID!
        name: String!
        nickname: String
        kind: Kind
      }

      type Person implements Entity {
        id: ID!
        name: String!
        nickname: String
        age: Int!
        relatives: [Person!]!
        favoritePet: Pet
      }

      type Query {
        person: Person
        entities: [Entity!]!
      }
    `);

    it('prints fragment definitions', () => {
      const fragment = `
        fragment Pets on Entity {
          name
          nickname
        }
      `;

      expect(print(fragment, schema)).toMatchSnapshot();
    });

    it('prints fragment spreads', () => {
      const fragment = `
        fragment Pets on Entity {
          ...PetKind
          ...PersonDetails
          ...EntityName
          id
        }
      `;

      expect(
        print(fragment, schema, {
          'subfolder/PetKind.graphql': 'fragment PetKind on Pet { kind }',
          'subfolder/another/PersonDetails.graphql': `
          fragment PersonDetails on Person {
            id
            name
            nickname
            favoritePet { name }
          }

          fragment EntityName on Entity { name }
        `,
        }),
      ).toMatchSnapshot();
    });

    it('prints inline fragments', () => {
      const fragment = `
        fragment Pets on Entity {
          ...on Pet {
            id
            kind
          }
          ...on Person {
            name
            nickname
            favoritePet { name }
          }
        }
      `;

      expect(print(fragment, schema)).toMatchSnapshot();
    });
  });

  describe('multiple operations', () => {
    const schema = buildSchema(`
      enum Kind {
        CAT
        DOG
      }

      type Pet {
        name: String!
        kind: Kind!
      }

      type Person {
        name: String!
        favoritePet: Pet
        pets: [Pet!]!
      }

      type Query {
        person: Person!
      }
    `);

    const document = `
      query Person {
        person {
          ...FavoritePet
        }
      }

      query Pet {
        person {
          pets {
            ...PetDetails
          }
        }
      }

      fragment PetDetails on Pet {
        kind
      }

      fragment FavoritePet on Person {
        name
        favoritePet {
          ...PetDetails
        }
      }
    `;

    expect(print(document, schema)).toMatchSnapshot();
  });

  describe('__typename', () => {
    const schema = buildSchema(`
      interface Entity {
        id: ID!
        name: String!
        nickname: String
        self: Entity!
      }

      enum Kind {
        DOG
        CAT
      }
    
      type Pet implements Entity {
        id: ID!
        name: String!
        nickname: String
        kind: Kind
        self: Pet!
      }

      type Person implements Entity {
        id: ID!
        name: String!
        nickname: String
        age: Int!
        relatives: [Person!]!
        favoritePet: Pet
        self: Entity!
      }

      union PetOrPerson = Pet | Person

      type Query {
        person: Person
        entities: [Entity!]!
        random: PetOrPerson
      }
    `);

    it('does not duplicate an explicit __typename field', () => {
      expect(
        print(
          `
        query Profile {
          person {
            __typename
          }
        }
      `,
          schema,
        ),
      ).toMatchSnapshot();
    });

    it('adds a __typename field when the responseName is unique', () => {
      expect(
        print(
          `
        query Profile {
          person {
            type: __typename
          }
        }
      `,
          schema,
        ),
      ).toMatchSnapshot();
    });

    it('does not add a typename when the option is set to false', () => {
      expect(
        print(
          `
        query Profile {
          person {
            name
          }
        }
      `,
          schema,
          {},
          {addTypename: false},
        ),
      ).toMatchSnapshot();
    });

    it('adds an explicit typename when the option is set to false', () => {
      expect(
        print(
          `
        query Profile {
          person {
            __typename
          }
        }
      `,
          schema,
          {},
          {addTypename: false},
        ),
      ).toMatchSnapshot();
    });

    it('prints a correct typename for interfaces', () => {
      expect(
        print(
          `
        fragment MyEntity on Entity {
          name
        }
      `,
          schema,
        ),
      ).toMatchSnapshot();

      expect(
        print(
          `
        query Profile {
          person {
            self {
              __typename
            }
          }
        }
      `,
          schema,
        ),
      ).toMatchSnapshot();

      expect(
        print(
          `
        query PetProfile {
          entities {
            ...on Pet {
              self {
                __typename
              }
            }
          }
        }
      `,
          schema,
        ),
      ).toMatchSnapshot();

      expect(
        print(
          `
        query RandomProfile {
          random {
            ...on Person {
              name
            }

            ...on Pet {
              name
            }
          }
        }
      `,
          schema,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('errors', () => {
    it('throws an error when referencing a missing type', () => {
      const schema = buildSchema(`
        enum Kind {
          DOG
          CAT
        }
      
        type Pet {
          id: ID!
          name: String!
          nickname: String
          kind: Kind
        }

        type Query {
          pet(kind: Kind!): Pet
        }
      `);

      expect(() =>
        print(
          `
        query RandomPet($kind: Kindd!) {
          pet(kind: $kind) {
            name
          }
        }
      `,
          schema,
        ),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

function print(
  documentString: string,
  schema: GraphQLSchema,
  fragments: {[key: string]: string} = {},
  options: Partial<Options> = {},
) {
  const finalOptions = {
    addTypename: true,
    ...options,
  };
  const fileName = 'MyOperation.graphql';
  const fragmentDocuments = Object.keys(fragments).map((key) =>
    parse(new Source(fragments[key], key)),
  );
  const document = parse(new Source(documentString, fileName));
  const ast = compile(schema, concatAST([document, ...fragmentDocuments]));
  const file = {
    path: fileName,
    operations: Object.keys(ast.operations)
      .map((key) => ast.operations[key])
      .filter((operation) => operation.filePath === fileName),
    fragments: Object.keys(ast.fragments)
      .map((key) => ast.fragments[key])
      .filter((fragment) => fragment.filePath === fileName),
  };
  return printFile(file, ast, finalOptions);
}
