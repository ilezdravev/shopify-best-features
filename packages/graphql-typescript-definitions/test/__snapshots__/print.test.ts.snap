// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export namespace Schema {
  export type Kind = 'CAT' | 'DOG';
}

export interface PersonQueryData {
  person: PersonQueryData.Person,
}

export namespace PersonQueryData {
  export interface Pet extends PetDetailsFragmentData {
    __typename: 'Pet',
    kind: Schema.Kind,
  }

  export interface Person extends FavoritePetFragmentData {
    __typename: 'Person',
    name: string,
    favoritePet?: PersonQueryData.Pet | null,
  }
}

export interface PetQueryData {
  person: PetQueryData.Person,
}

export namespace PetQueryData {
  export interface Pet extends PetDetailsFragmentData {
    __typename: 'Pet',
    kind: Schema.Kind,
  }

  export interface Person {
    __typename: 'Person',
    pets: PetQueryData.Pet[],
  }
}

export interface PetDetailsFragmentData {
  __typename: 'Pet',
  kind: Schema.Kind,
}

export interface FavoritePetFragmentData {
  __typename: 'Person',
  name: string,
  favoritePet?: FavoritePetFragmentData.Pet | null,
}

export namespace FavoritePetFragmentData {
  export interface Pet extends PetDetailsFragmentData {
    __typename: 'Pet',
    kind: Schema.Kind,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename adds a __typename field when the responseName is unique 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProfileQueryData {
  person?: ProfileQueryData.Person | null,
}

export namespace ProfileQueryData {
  export interface Person {
    __typename: 'Person',
    type: 'Person',
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename adds an explicit typename when the option is set to false 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProfileQueryData {
  person?: ProfileQueryData.Person | null,
}

export namespace ProfileQueryData {
  export interface Person {
    __typename: 'Person',
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename does not add a typename when the option is set to false 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProfileQueryData {
  person?: ProfileQueryData.Person | null,
}

export namespace ProfileQueryData {
  export interface Person {
    name: string,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename does not duplicate an explicit __typename field 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProfileQueryData {
  person?: ProfileQueryData.Person | null,
}

export namespace ProfileQueryData {
  export interface Person {
    __typename: 'Person',
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename prints a correct typename for interfaces 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface MyEntityFragmentData {
  __typename: 'Person' | 'Pet',
  name: string,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename prints a correct typename for interfaces 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProfileQueryData {
  person?: ProfileQueryData.Person | null,
}

export namespace ProfileQueryData {
  export interface Entity {
    __typename: 'Person' | 'Pet',
  }

  export interface Person {
    __typename: 'Person',
    self: ProfileQueryData.Entity,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename prints a correct typename for interfaces 3`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface PetProfileQueryData {
  entities: PetProfileQueryData.Entity[],
}

export namespace PetProfileQueryData {
  export interface Pet {
    __typename: 'Pet',
  }

  export interface Entity {
    __typename: 'Person' | 'Pet',
    self?: PetProfileQueryData.Pet | null,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() __typename prints a correct typename for interfaces 4`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface RandomProfileQueryData {
  random?: RandomProfileQueryData.PetOrPerson | null,
}

export namespace RandomProfileQueryData {
  export interface PetOrPerson {
    __typename: 'Pet' | 'Person',
    name?: string | null,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() custom scalars declares them as strings 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export namespace Schema {
  export type DateTime = string;
}

export interface DateQueryData {
  date: Schema.DateTime,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() custom scalars declares them as strings 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export namespace Schema {
  export type DateTime = string;
}

export interface DateQueryData {
  date?: Schema.DateTime | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() enum types prints the enum type and its type declaration 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export namespace Schema {
  export type Episode = 'ONE' | 'TWO' | 'THREE';
}

export interface WhichEpisodeQueryData {
  episode: Schema.Episode,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() enum types prints the enum type and its type declaration 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export namespace Schema {
  export type Episode = 'ONE' | 'TWO' | 'THREE';
}

export interface WhichEpisodeQueryData {
  episode?: Schema.Episode | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() errors throws an error when referencing a missing type 1`] = `"An invalid type was provided for key 'kind'. This probably means there is a typo in your document or that one of your input types has been renamed."`;

exports[`printFile() fragments prints fragment definitions 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface PetsFragmentData {
  __typename: 'Person' | 'Pet',
  name: string,
  nickname?: string | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() fragments prints fragment spreads 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

import {PetKindFragmentData} from './subfolder/PetKind.graphql';
import {PersonDetailsFragmentData, EntityNameFragmentData} from './subfolder/another/PersonDetails.graphql';

type NullableFragment<T> = {
  [P in keyof T]?: T[P] | null
}
export namespace Schema {
  export type Kind = 'DOG' | 'CAT';
}

export interface PetsFragmentData extends NullableFragment<PetKindFragmentData>, NullableFragment<PersonDetailsFragmentData>, EntityNameFragmentData {
  __typename: 'Person' | 'Pet',
  name: string,
  id: string,
  kind?: Schema.Kind | null,
  nickname?: string | null,
  favoritePet?: PetsFragmentData.Pet | null,
}

export namespace PetsFragmentData {
  export interface Pet {
    __typename: 'Pet',
    name: string,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() fragments prints inline fragments 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export namespace Schema {
  export type Kind = 'DOG' | 'CAT';
}

export interface PetsFragmentData {
  __typename: 'Person' | 'Pet',
  id: string,
  kind?: Schema.Kind | null,
  name: string,
  nickname?: string | null,
  favoritePet?: PetsFragmentData.Pet | null,
}

export namespace PetsFragmentData {
  export interface Pet {
    __typename: 'Pet',
    name: string,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() list types correctly wraps types that are potentially null 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NicknamesQueryData {
  nicknames: string[],
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() list types correctly wraps types that are potentially null 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NicknamesQueryData {
  nicknames: (string | null)[],
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() list types correctly wraps types that are potentially null 3`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NicknamesQueryData {
  nicknames?: string[] | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() list types correctly wraps types that are potentially null 4`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NicknamesQueryData {
  nicknames?: (string | null)[] | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() list types wraps nested lists 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NicknamesQueryData {
  nicknames: (string[] | null)[],
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() list types wraps nested lists 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NicknamesQueryData {
  nicknames?: ((string | null)[] | null)[] | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() mutations prints mutations 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface RenameMutationData {
  rename: RenameMutationData.Person,
}

export namespace RenameMutationData {
  export interface Variables {
    name: string,
  }

  export interface Person {
    __typename: 'Person',
    name: string,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() object types fragment spreads prints a partial fragment type when the type restriction is not guaranteed 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

import {EntityNameFragmentData} from './subfolder/MyFragment.graphql';

type NullableFragment<T> = {
  [P in keyof T]?: T[P] | null
}
export interface EntitiesQueryData {
  entities: EntitiesQueryData.Entity[],
}

export namespace EntitiesQueryData {
  export interface Entity extends NullableFragment<EntityNameFragmentData> {
    __typename: 'Person' | 'Pet',
    name?: string | null,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() object types fragment spreads prints them as union types when the type restriction is guaranteed 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

import {EntityNameFragmentData} from './subfolder/MyFragment.graphql';

export interface EntitiesQueryData {
  entities: EntitiesQueryData.Entity[],
}

export namespace EntitiesQueryData {
  export interface Entity extends EntityNameFragmentData {
    __typename: 'Person' | 'Pet',
    name: string,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() object types inline fragments forces fields to be nullable when the type restriction is not guaranteed 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface EntitiesQueryData {
  entities: EntitiesQueryData.Entity[],
}

export namespace EntitiesQueryData {
  export interface Entity {
    __typename: 'Person' | 'Pet',
    name?: string | null,
    nickname?: string | null,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() object types inline fragments treats them as their original nullability when they are guaranteed 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface EntitiesQueryData {
  entities: EntitiesQueryData.Entity[],
}

export namespace EntitiesQueryData {
  export interface Entity {
    __typename: 'Person' | 'Pet',
    name: string,
    nickname?: string | null,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() object types objects fields nests them inline in the original object 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProfileQueryData {
  person?: ProfileQueryData.PersonPerson | null,
}

export namespace ProfileQueryData {
  export interface Pet {
    __typename: 'Pet',
    name: string,
  }

  export interface PersonRelativesPerson {
    __typename: 'Person',
    name: string,
    favoritePet?: ProfileQueryData.Pet | null,
  }

  export interface PersonPerson {
    __typename: 'Person',
    name: string,
    age: number,
    relatives: ProfileQueryData.PersonRelativesPerson[],
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints a boolean type 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface MarriedQueryData {
  married: boolean,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints a boolean type 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface MarriedQueryData {
  married?: boolean | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints a float type 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface AgeQueryData {
  age: number,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints a float type 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface AgeQueryData {
  age?: number | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints a string type 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NameQueryData {
  name: string,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints a string type 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface NameQueryData {
  name?: string | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints an ID type 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProductIDQueryData {
  id: string,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints an ID type 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface ProductIDQueryData {
  id?: string | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints an integer type 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface AgeQueryData {
  age: number,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() primitive types prints an integer type 2`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export interface AgeQueryData {
  age?: number | null,
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;

exports[`printFile() variables prints out a variable interface and all associated types 1`] = `
"// This file was generated and should not be edited.
/* tslint:disable */

import {DocumentNode} from 'graphql';

export namespace Schema {
  export interface Language {
    code?: string | null,
  }

  export type DateTime = string;

  export type Style = 'FULL' | 'NICK';

  export interface NameInput {
    style: Schema.Style,
    date?: Schema.DateTime | null,
    excited: boolean,
    language?: Schema.Language | null,
  }
}

export interface NameQueryData {
  name: string,
}

export namespace NameQueryData {
  export interface Variables {
    myInput?: Schema.NameInput | null,
  }
}

declare const document: DocumentNode;
export default document;

/* tslint:enable */
"
`;
